amads/melody/narmour.py

```diff
@@ -0,0 +1,208 @@
+            real_interval = abs(pitches[i] - pitches[i-1])
+            # Smaller difference = higher expectancy
+            expectations[i] = 1.0 / (1.0 + abs(impl_interval - real_interval))
+        return expectations
+    
+    def closure(pitches: np.ndarray) -> np.ndarray:
+        """Calculate closure principle."""
+        expectations = np.full(len(pitches), np.nan)
+        for i in range(2, len(pitches)):
+            impl_interval = abs(pitches[i-1] - pitches[i-2])
+            real_interval = abs(pitches[i] - pitches[i-1])
+            # Closure occurs when realized interval is smaller
+            expectations[i] = float(real_interval < impl_interval)
+        return expectations
+    
+    def proximity(pitches: np.ndarray) -> np.ndarray:
+        """Calculate revised proximity principle."""
+        expectations = np.full(len(pitches), np.nan)
+        for i in range(2, len(pitches)):
+            # Smaller intervals are more expected
+            interval = abs(pitches[i] - pitches[i-1])
+            expectations[i] = 1.0 / (1.0 + interval)
+        return expectations
+    
+    def consonance(pitches: np.ndarray) -> np.ndarray:
+        """Calculate consonance principle."""
+        # Consonance ratings from Krumhansl (1995)
+        CONSONANCE_RATINGS = {
+            0: 1.0,   # unison
+            3: 0.6,   # minor third
+            4: 0.8,   # major third
+            5: 0.7,   # perfect fourth
+            7: 0.9,   # perfect fifth
+            8: 0.6,   # minor sixth
+            9: 0.7,   # major sixth
+            12: 0.9   # octave
+        }
+        
+        expectations = np.full(len(pitches), np.nan)
+        for i in range(2, len(pitches)):
+            interval = abs(pitches[i] - pitches[i-1]) % 12
+            expectations[i] = CONSONANCE_RATINGS.get(interval, 0.1)
+        return expectations
+
+    # Flatten and collapse the score into a single sequence of notes
+    flattened_score = score.flatten(collapse=True)
+    notes = list(flattened_score.find_all(Note))
+    
+    # Handle empty scores or sequences too short for analysis
+    if len(notes) < 3:
+        return np.array([])
+    
+    # Extract pitch values
+    pitches = np.array([note.keynum for note in notes])
+    
+    # Calculate expectations based on selected principle
+    principle_functions = {
+        'rd': registral_direction,
+        'rr': registral_return,
+        'id': intervallic_difference,
+        'cl': closure,
+        'pr': proximity,
+        'co': consonance
+    }
+    
+    return principle_functions[principle](pitches)
+
+
+if __name__ == "__main__":
+    # Test case 1: Simple ascending melody
+    print("\nTest Case 1: Simple ascending melody")
+    score = Score()
+    part = Part()
+    notes = [
+        Note(pitch=60),  # C4
+        Note(pitch=64),  # E4
+        Note(pitch=67),  # G4
+        Note(pitch=72)   # C5
+    ]
+    for note in notes:
+        part.append(note)
+    score.append(part)
```

Comment:
A lot of these kinds of passages can now be simplified using the new `Score.get_notes` method.